### DISCUSSION TITLE
Realtime column selectors and filtering?

### DISCUSSION THREAD

**Discussion by benallfree** [pocketbase/pocketbase#6141] [Q&A]
Is there a way to constrain realtime results to specific columns and only send updates if the values of those columns have changed?

Edit: I think I see it here: https://github.com/pocketbase/pocketbase/blob/08f2190ad1e732e12d5e31ff8b59a116b4ef2af1/apis/realtime.go#L568

It looks like it can support the query params from `/api/collections/collectionIdOrName/records`, is that right?

**Comment by benallfree** [comment]
Followup question: I don't see a reference to `e.Record.Original` in realtime, so it seems it will broadcast even if none of the `fields` specified have changed value. 

Are you interested in a PR that will suppress `realtimeBroadcastRecord` if all the requested fields still match the original?

**Comment by ganigeorgiev** [comment]
> It looks like it can support the query params from /api/collections/collectionIdOrName/records, is that right?

Yes.

> Are you interested in a PR that will suppress realtimeBroadcastRecord if all the requested fields still match the original?

It is not clear what you have in mind and more importantly "why" what you want needs to be added in the first place? If you want to change the behavior of the `fields` parameter, then No, there are no plans for this.

**Reply by benallfree** [reply]
It looks like the current implementation will broadcast updates even if the final fields in [cleanRecord](https://github.com/pocketbase/pocketbase/blob/08f2190ad1e732e12d5e31ff8b59a116b4ef2af1/apis/realtime.go#L507) are unchanged from  the values in `record.Original()`.

**Reply by ganigeorgiev** [reply]
Yes and this is intentional because `fields` only select/pick record fields, aka. it should work the same way as if you are retrieving the record with a regular `getOne` request. Note that `fields` technically also supports modifiers, e.g. `title:something` (_this is not documented in details at the moment because I'm still not sure how to properly expose it but currently the Dashboard for example uses the `:excerpt` modifier for the records list to avoid rendering large `json` or `editor` field values_).

Again, please elaborate why this is needed in the first place. We can eventually introduce a dedicated parameter to listen for specific changed fields, but is this really worth the extra complexity and performance penalty that it will introduce?

**Reply by benallfree** [reply]
PocketHost uses a central PocketBase instance (the "mothership") to track the status and stats of individual instances (e.g., starting, stopping, running, idle). Each PocketHost edge node maintains a real-time connection to the mothership, caching instance states for efficient operation. However, the edge nodes only need to be notified when specific fields change. 

Currently, they receive real-time updates for all field changes, even when the updates are irrelevant to their needs. This creates unnecessary communication overhead.

While splitting collections to isolate relevant fields is one option, a better approach would be to improve the real-time system to send updates only when the requested fields have actually changed. This would reduce excess chatter and make the system more efficient.

(*As a side note, I think I can provide some real-world data to improve the realtime system in other ways, too. A user's app has gone viral and it seems to be bottlenecking somewhere in the realtime code. Queries go from 10ms to 2000ms+ as realtime connections grow into the thousands. Disabling realtime connections returns the system to normal.*)

**Reply by ganigeorgiev** [reply]
> As a side note, I think I can provide some real-world data to improve the realtime system in other ways, too. A user's app has gone viral and it seems to be bottlenecking somewhere in the realtime code. Queries go from 10ms to 2000ms+ as realtime connections grow into the thousands. Disabling realtime connections returns the system to normal.

If you think that something can be improved in the realtime handling, please feel free to open a PR and I'll try to review it, but the above doesn't really help and I hate discussing performance when there is nothing concrete to work on. There are at least dozens opened issues about query performance and most users just abandon their requests (https://github.com/pocketbase/pocketbase/discussions/5879, https://github.com/pocketbase/pocketbase/discussions/5847) and I'm tired by this.

**Reply by benallfree** [reply]
Do you want a PR for suppressing unchanged updates?

**Reply by ganigeorgiev** [reply]
If you can show that this will improve the performance (_the record diff values change could be slower than just returning the record because it will have to be performed for each subscription that has the specific param_), then I'll accept it as a new parameter, but I'm not willing to complicate further the internals without any data or good reasons to do so.

Regarding "_Disabling realtime connections returns the system to normal_" this could be caused by many things, including OS level limitations or throttling on reverse proxy level, especially in your case with running several PocketBase instances on the same machine behind a reverse proxy. I'm willing to investigate it or review a PR for that if you can isolate it but I'm not willing to invest my time on a wild goose chase.
